### PR TITLE
NT-1531: Android Refactor RewardUtils Java to Kotlin

### PR DESCRIPTION
**Pull request by leighdouglas**
# 📲 What

Convert the RewardUtils file and associated tests from Java to Kotlin

# 🤔 Why

Apart of our efforts to address technical debt. 

# Story 📖

[NT-1531: Android Refactor RewardUtils Java to Kotlin](https://kickstarter.atlassian.net/browse/NT-1531?atlOrigin=eyJpIjoiMzFkYTQ0NmQxZWU5NGNiOTljMWZkNzBhY2E4MTlkYzMiLCJwIjoiaiJ9)

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+    fun isAvailable(project: Project, reward: Reward): Boolean {
+        return project.isLive && !isLimitReached(reward) && !isExpired(reward)
+    }
+
+    /**
+     * Returns `true` if the reward has expired.
+     */
+    fun isExpired(reward: Reward): Boolean {
+        return isTimeLimited(reward) && reward.endsAt()!!.isBeforeNow
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:32]: Use un here the unwrap technique with elvis operator, do not force an option with the !! operator 
`reward.endsAt()?.let { it.isBeforeNow } ?: false`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+import org.joda.time.Duration
+import kotlin.math.floor
+import kotlin.math.max
+
+object RewardUtils {
+
+    /**
+     * Returns `true` if the reward has backers, `false` otherwise.
+     */
+    fun hasBackers(reward: Reward): Boolean {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:20]: you can transform this in an inline function, it just have one condition.
`fun hasBackers(reward: Reward) = IntegerUtils.isNonZero(reward.backersCount())`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    fun isLimitReached(reward: Reward): Boolean {
+        val remaining = reward.remaining()
+        return reward.limit() != null && remaining != null && remaining <= 0
+    }
+
+    /**
+     * Returns `true` if the reward is considered the 'non-reward' option, i.e. the reward is the option
+     * backers select when they want to pledge to a project without selecting a particular reward.
+     */
+    fun isNoReward(reward: Reward): Boolean {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:62]: inline function here as well 
`fun isNoReward(reward: Reward) = reward.id() == 0L`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        return reward.id() == 0L
+    }
+
+    /**
+     * Returns `true` if the reward is a specific reward for a project, i.e. it is not the 'no-reward' option.
+     */
+    fun isReward(reward: Reward): Boolean {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:69]: same here, inline function
`fun isReward(reward: Reward) = !isNoReward(reward)`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        return (reward.shippingPreferenceType() == Reward.ShippingPreference.NONE || reward.shippingPreferenceType() == Reward.ShippingPreference.NOSHIPPING ||
+                isDigitalV1) && !isShippable(reward)
+    }
+
+    /**
+     * Returns `true` if the reward has a valid expiration date.
+     */
+    fun isTimeLimited(reward: Reward): Boolean {
+//         TODO: 2019-06-14 remove epoch check after Garrow fixes `current` bug in backend
+        return reward.endsAt() != null && !DateTimeUtils.isEpoch(reward.endsAt()!!)
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:100]: unwrapping with .let and elvis operator, do not force optionals 
`/**
     * Returns `true` if the reward has a valid expiration date.
     * // TODO: 2019-06-14 remove epoch check after Garrow fixes `current` bug in backend
     */
    fun isTimeLimited(reward: Reward) = reward.endsAt()?.let { !DateTimeUtils.isEpoch(it) } ?: false`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    /**
+     * Returns unit of time remaining in a readable string, e.g. `days to go`, `hours to go`.
+     */
+    fun deadlineCountdownDetail(reward: Reward, context: Context,
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:106]: inline function 
` fun deadlineCountdownDetail(reward: Reward, context: Context, ksString: KSString) = 
            ksString.format(context.getString(R.string.discovery_baseball_card_time_left_to_go),
                "time_left", deadlineCountdownUnit(reward, context))`

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * Returns the most appropriate unit for the time remaining until the reward
+     * reaches its deadline.
+     *
+     * @param context an Android context.
+     * @return the String unit.
+     */
+    fun deadlineCountdownUnit(reward: Reward, context: Context): String {
+        val seconds = timeInSecondsUntilDeadline(reward)
+        return when {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:122]: Great! the 'when' expression is hard to use great job here :)

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            return null
+        }
+
+        return when (shippingType) {
+            Reward.SHIPPING_TYPE_ANYWHERE -> Pair.create(R.string.Ships_worldwide, null)
+            Reward.SHIPPING_TYPE_MULTIPLE_LOCATIONS -> Pair.create(R.string.Limited_shipping, null)
+            Reward.SHIPPING_TYPE_SINGLE_LOCATION -> {
+                val location = reward.shippingSingleLocation()
+                return if (ObjectUtils.isNotNull(location)) {
+                    Pair.create(R.string.location_name_only, location!!.localizedName())
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:150]: same here unwrap, do not force optionals

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+            else -> null
+        }
+    }
+
+    /**
+     * Returns time until reward reaches deadline in seconds, or 0 if the
+     * reward has already finished.
+     */
+    //TODO
+    fun timeInSecondsUntilDeadline(reward: Reward): Long {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:164]: inline function 
`    fun timeInSecondsUntilDeadline(reward: Reward) =
            max(0L, Duration(DateTime(), reward.endsAt()).standardSeconds)`

**Review by Arkariang — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt (hunk 7 above)
Comment (leighdouglas) [reply app/src/main/java/com/kickstarter/libs/utils/RewardUtils.kt:109]: Thank you!

**Review by jgsamudio — APPROVED** [review]
Great Work! 👍

File: app/src/test/java/com/kickstarter/libs/utils/RewardUtilsTest.kt [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+    private val DAYS_TO_GO = "days to go"
+    private val HOURS_TO_GO = "hours to go"
+    private val SECS_TO_GO = "secs to go"
+    private val DAYS = "days"
+    private val HOURS = "hours"
+    private val MINS = "mins"
+    private val SECS = "secs"
+    private val COUNTRY_NIGERIA = "Nigeria"
+
+    private lateinit var context : Context
```
Comment (Arkariang) [inline app/src/test/java/com/kickstarter/libs/utils/RewardUtilsTest.kt:46]: Great work here! lateinit vars are really useful 👍

**Comment by Arkariang** [comment]
I really love how RewardUtilsTest.kt looks like, so tidy and neat :sparkling_heart: :heart_eyes_cat:

**Review by Arkariang — APPROVED** [review]

**Pull request merged**